File: 프로그래머스/1/42889. 실패율/실패율.py
from collections import Counter, defaultdict
def solution(N, stages):
    answer = []
    failure = {}
    total = len(stages)
    d = defaultdict(int, Counter(stages))

    for i in range(1, N+1):
        if total == 0:
            failure[i] = 0
        else:
            failure[i] = d[i] / total
            total -= d[i]
    answer = [k for k, v in sorted(failure.items(), key=lambda x:(-x[1],x[0]))]
        
    return answer

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        
        int total = stages.length;
        Map<Integer,Integer> map = new HashMap<>(); 
        for(int s: stages) {
            map.put(s, map.getOrDefault(s,0)+1);
        }
        
        Map<Integer,Double> failure = new HashMap<>(); // index,failure
        for(int i=1;i<N+1;i++) {
            if(map.containsKey(i)) {
                failure.put(i, (double)map.get(i)/total);
                total -= map.get(i);
            } else {
                failure.put(i, 0.0);
            }
        }
        
        List<Map.Entry<Integer,Double>> entry = new ArrayList<>(failure.entrySet());
        entry.sort((a,b) -> {
            if(a.getValue()!=b.getValue()) {
                return Double.compare(b.getValue(),a.getValue());
            }
            return a.getKey() - b.getKey();
        });
        
        int idx = 0;
        for(Map.Entry<Integer,Double> en: entry) {
            answer[idx++] = en.getKey();
        }
        
        return answer;
    }
}